src/cmd/watch/poller.rs

```diff
@@ -0,0 +1,386 @@
+//! A simple poll-based filesystem watcher.
+//!
+//! This exists because the native change notifications have historically had
+//! lots of problems. Various operating systems and different filesystems have
+//! had problems correctly reporting changes.
+
+use ignore::gitignore::Gitignore;
+use mdbook::MDBook;
+use pathdiff::diff_paths;
+use std::collections::HashMap;
+use std::fs::FileType;
+use std::path::{Path, PathBuf};
+use std::time::{Duration, Instant, SystemTime};
+use walkdir::WalkDir;
+
+/// Calls the closure when a book source file is changed, blocking indefinitely.
+pub fn rebuild_on_change(
+    book_dir: &Path,
+    update_config: &dyn Fn(&mut MDBook),
+    post_build: &dyn Fn(),
+) {
+    let mut book = MDBook::load(book_dir).unwrap_or_else(|e| {
+        error!("failed to load book: {e}");
+        std::process::exit(1);
+    });
+
+    let mut watcher = Watcher::new(book_dir);
+
+    info!("Watching for changes...");
+    // Scan once to initialize the starting point.
+    watcher.set_roots(&book);
+    watcher.scan();
+
+    // Track average scan time, to help investigate if the poller is taking
+    // undesirably long. This is not a rigorous benchmark, just a rough
+    // estimate.
+    const AVG_SIZE: usize = 60;
```

Comment:
I'm a little concerned that would cause too many false positives. For example, just a temporary hiccup, or the user puts their machine to sleep, or whatever. I think we can add more here in the future if people report having issues.